installer/src/pinephone.rs

```diff
@@ -0,0 +1,276 @@
+        .send_at_command(&adb_enable.to_string())
+        .await
+        .context("Failed to send enable adb command.")?;
+    let resp = String::from_utf8_lossy(&start_adb);
+    if !resp.contains("\r\nOK\r\n") {
+        bail!("Received unexpected response: {0}", resp);
+    }
+
+    Ok(())
+}
+
+/// Stop the adb daemon on the Quectel modem.
+pub async fn stop_adb() -> Result<()> {
+    let tty = serial_interface()?.unwrap();
+    let adb_disable = Qusbcfg::default();
+    let stop_adb = tty
+        .send_at_command(&adb_disable.to_string())
+        .await
+        .context("Failed to disable adb.")?;
+    let resp = String::from_utf8_lossy(&stop_adb);
+    if !resp.contains("\r\nOK\r\n") {
+        bail!("Received unexpected response: {0}", resp);
+    }
+    Ok(())
+}
+
+trait Install {
+    fn run_command(&mut self, command: &[&str], expected_output: &str) -> Result<()>;
+    fn install_file(&mut self, dest: &str, payload: &[u8]) -> Result<()>;
+}
+
+impl Install for ADBUSBDevice {
+    /// Run an adb shell command, append '; echo exit code $?' to the command and verify its output.
+    fn run_command(&mut self, command: &[&str], expected_output: &str) -> Result<()> {
+        let mut buf = Vec::<u8>::new();
+        let mut cmd = Vec::<&str>::new();
+        cmd.extend_from_slice(command);
+        cmd.extend_from_slice(&[";", "echo", "exit code $?"]);
+        self.shell_command(&cmd, &mut buf)?;
+        let output = String::from_utf8_lossy(&buf);
+        if !output.contains(expected_output) {
+            bail!("{expected_output:?} not found in: {output}");
+        }
+        Ok(())
+    }
+
+    /// Transfer a file to the modem's filesystem with adb push.
+    /// Validates the file sends successfully to /tmp before overwriting the destination.
+    fn install_file(&mut self, dest: &str, mut payload: &[u8]) -> Result<()> {
+        echo!("Sending file {dest} ... ");
+        let file_name = Path::new(dest)
+            .file_name()
+            .ok_or_else(|| anyhow!("{dest} does not have a file name"))?
+            .to_str()
+            .ok_or_else(|| anyhow!("{dest}'s file name is not UTF8"))?
+            .to_owned();
+        let push_tmp_path = format!("/tmp/{file_name}");
+        let file_hash = md5_compute(payload);
+        self.push(&mut payload, &push_tmp_path)?;
+        self.run_command(&["md5sum", &push_tmp_path], &format!("{file_hash:x}"))?;
+        self.run_command(&["mv", &push_tmp_path, dest], "exit code 0")?;
+        println!("ok");
+        Ok(())
+    }
+}
+
+/// Claim the modem's USB interface for sending AT commands.
+fn serial_interface() -> Result<Option<Interface>> {
+    if let Some(device) = open_usb_device(USB_VENDOR_ID, USB_PRODUCT_ID)? {
+        let interface = device
+            .detach_and_claim_interface(USB_INTERFACE_NUMBER)
+            .context("detach_and_claim_interface({USB_INTERFACE_NUMBER}) failed")?;
+        return Ok(Some(interface));
+    }
+    Ok(None)
+}
+
+trait AT {
```

Comment:
this seems like you were trying to abstract something away between pinephone and orbic, but now there's only one impl of this trait within this module, so I'd suggest removing it again